src/checkpointTrie.js

```diff
@@ -111,60 +118,50 @@ module.exports = class CheckpointTrie extends BaseTrie {
   createScratchReadStream (scratch) {
     const trie = this.copy()
     scratch = scratch || this._scratch
+    scratch = new DB(scratch._db)
     // Only read from the scratch
-    trie._getDBs = [scratch]
+    trie.db = scratch
     trie._scratch = scratch
     return new ScratchReadStream(trie)
   }
 
+  putRaw (key, val, cb) {
+    if (this.isCheckpoint) return this._overridePutRaw(key, val, cb)
+    this.db.put(key, val, cb)
+  }
+
   /**
    * Puts kv-pair directly to db, ignoring checkpoints.
    * @private
    */
   _overridePutRaw (key, val, cb) {
-    const dbPut = (db, cb2) => {
-      db.put(key, val, {
-        keyEncoding: 'binary',
-        valueEncoding: 'binary'
-      }, cb2)
-    }
-    async.each(this.__putDBs, dbPut, cb)
+    this._mainDB.put(key, val, cb)
   }
 
   /**
    * Enter into checkpoint mode.
    * @private
    */
   _enterCpMode () {
-    this._scratch = level()
-    this._getDBs = [this._scratch].concat(this._getDBs)
-    this.__putDBs = this._putDBs
-    this._putDBs = [this._scratch]
-    this._putRaw = this.putRaw
-    this.putRaw = this._overridePutRaw
+    this._scratch = new ScratchDB(this._mainDB)
+    this.db = this._scratch
   }
 
   /**
    * Exit from checkpoint mode.
    * @private
    */
   _exitCpMode (commitState, cb) {
-    var scratch = this._scratch
+    const scratch = this._scratch
     this._scratch = null
-    this._getDBs = this._getDBs.slice(1)
-    this._putDBs = this.__putDBs
-    this.putRaw = this._putRaw
+    this.db = this._mainDB
 
-    const flushScratch = (db, cb) => {
+    if (commitState) {
       this.createScratchReadStream(scratch)
-        .pipe(WriteStream(db))
+        .pipe(WriteStream(this.db))
         .on('close', cb)
-    }
-
-    if (commitState) {
-      async.map(this._putDBs, flushScratch, cb)
     } else {
-      cb()
+      async.nextTick(cb)
```

Comment:
Is this `async.nextTick()` actually fixing this proxy method thing?

Reply:
Generally went through the whole new checkpoint process and checked if this works and is consistent, looks good.

Reply:
The `nextTick` is there mainly to prevent ["releasing Zalgo"](https://blog.izs.me/2013/08/designing-apis-for-asynchrony) (see https://github.com/ethereumjs/ethereumjs-vm/issues/421).

Reply:
Ah, thanks.